src/accelerate/inference.py

```diff
@@ -0,0 +1,155 @@
+
+    Users can pass in custom `num_chunks` as an optional hyper-parameter. By default will use
+    `AcceleratorState.num_processes`
+    """
+    # We need to annotate the split points in the model for PiPPy
+    state = PartialState()
+    annotate_split_points(model, {split_point: PipeSplitWrapper.SplitPoint.BEGINNING for split_point in split_points})
+    found_batch_size = find_pippy_batch_size(args, kwargs)
+    if found_batch_size != num_chunks:
+        args = pad_input_tensors(args, found_batch_size, num_chunks)
+        kwargs = pad_input_tensors(kwargs, found_batch_size, num_chunks)
+    pipe = Pipe.from_tracing(model, num_chunks=num_chunks, example_args=args, example_kwargs=kwargs)
+    stage = PipelineStage(pipe, state.local_process_index, device=state.device)
+
+    return stage
+
+
+def pippy_forward(forward, *args, **kwargs):
+    state = PartialState()
+    output = None
+
+    if state.num_processes == 1:
+        output = forward(*args, **kwargs)
+    elif state.is_local_main_process:
+        found_batch_size = find_pippy_batch_size(args, kwargs)
+        if found_batch_size is None:
+            raise ValueError("Could not find batch size from args or kwargs")
+        else:
+            if found_batch_size != state.num_processes:
+                args = pad_input_tensors(args, found_batch_size, state.num_processes)
+                kwargs = pad_input_tensors(kwargs, found_batch_size, state.num_processes)
+        forward(*args, **kwargs)
+    elif state.is_last_process:
+        output = forward()
+    else:
+        forward()
+    return output
+
+
+def prepare_pippy(
+    model, split_points="auto", no_split_module_classes=None, example_args=(), example_kwargs={}, num_chunks=None
+):
+    """
+    Wraps `model` for PipelineParallelism
+
+    Args:
+        model (`torch.nn.Module`):
+            A model we want to split for pipeline-parallel inference
+        split_points (`str`, defaults to 'auto'):
+            How to generate the split points and chunk the model across each GPU. 'auto' will find the best balanced
+            split given any model.
+        no_split_module_classes (`List[str]`):
+            A list of class names for layers we don't want to be split.
+        example_args (tuple of `torch.Tensor`):
+            The expected inputs for the model that uses order-based inputs. Recommended to use this method if possible.
+        example_kwargs (dict of `torch.Tensor`)
+            The expected inputs for the model that uses dictionary-based inputs. This is a *highly* limiting structure
+            that requires the same keys be present at *all* inference calls. Not recommended unless the prior condition
+            is true for all cases.
+        num_chunks (`int`):
+            The number of different stages the Pipeline will have. By default it will assign one chunk per GPU, but
+            this can be tuned and played with. In general one should have num_chunks > num_gpus.
```

Comment:
Hi @kwen2501, could you help us to better understand what this arg do ? Is this the same arg as mentioned in this [doc](https://huggingface.co/docs/transformers/v4.15.0/parallelism#:~:text=PP%20introduces%20a,B0%2C2%2C%20B0%2C1%2C%20B0%2C0) ? Let's say we set a `num_chunks` of 4. It will split the data of `batch_size` n into 4 chunks of batch_size `n/4` ? And if we set `num_chunks=1`, we have the naive MP case. It that right ?

Reply:
That's right. Same as the definition in the doc you linked, i.e. how many microbatches we cut the input batch into.

Note that `num_chunks` is not necessarily the number of pipeline stages. For example, you can still have `num_chunks=1` but 4 pipeline stages. That 1 chunk would pass through the 4 stages in series. But since there are no more data chunks, a stage (GPU) would become idle once the chunk passes through it.

Reply:
That said, I agree with the comment that says:
> In general one should have num_chunks >= num_gpus

(maybe adding a "=" sign too)

Reply:
Thanks for clarifying @kwen2501 !